### PR TITLE
Context propagation across HTTP.

### LINKED ISSUE #4256: support for transport level context propagation
Between client-side and server-side, we should be able to provide a means for developers to provide an extensible mechanism to have certain context attributes propagated (presumably over headers) to the remote server over a Helidon any Http Transport.  @tomas-langer provided an example of this in an example he provided to some internal teams.  The enhancement request here is to turn that example into a 1st class, supported framework mechanism in Helidon.

Relates tangentially to #4016

### PR DESCRIPTION
**Pull request by tomas-langer**
Resolves #4256 for 2.x

Signed-off-by: Tomas Langer <tomas.langer@oracle.com>

### CODE REVIEW

File: tests/integration/webserver/context-propagation/src/test/java/io/helidon/tests/integration/webserver/context/propagation/ContextPropagationTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        assertThat(dto.getValue(), nullValue());
+        assertThat(dto.getValues(), nullValue());
+        assertThat(dto.getDefaultValue(), is("default-value"));
+        assertThat(dto.getDefaultValues(), is(arrayContaining("default-value1",
+                                                              "default-value2",
+                                                              "default-value3")));
+    }
+
+    @Test
+    void testAllCustom() {
```
Comment (cpotter1) [inline tests/integration/webserver/context-propagation/src/test/java/io/helidon/tests/integration/webserver/context/propagation/ContextPropagationTest.java:67]: would be nice to see a negative test case, where a particular value in the context is NOT propagated across.

File: webclient/context-propagation/pom.xml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        <dependency>
+            <groupId>io.helidon.common</groupId>
+            <artifactId>helidon-common-context</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.webclient</groupId>
+            <artifactId>helidon-webclient</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.config</groupId>
```
Comment (cpotter1) [inline webclient/context-propagation/pom.xml:43]: for my own understanding, why isn't this a normal compile time dep?

File: webserver/context-propagation/src/main/java/io/helidon/webserver/context/propagation/PropagationRecord.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                classifier = headerName;
+            } else if (headerName == null) {
+                headerName = classifier;
+            }
+            if (headerName == null) {
+                throw new IllegalArgumentException("Either header name or classifier must be configured");
+            }
+
+            if (isArray) {
+                return new ArrayRecord(classifier, headerName, defaultValue);
```
Comment (cpotter1) [inline webserver/context-propagation/src/main/java/io/helidon/webserver/context/propagation/PropagationRecord.java:72]: any considerations for record size limit / thresholds?

**Review by cpotter1 — APPROVED** [review]

File: webclient/context-propagation/pom.xml (hunk 2 above)
Comment (tomas-langer) [reply webclient/context-propagation/pom.xml:43]: Because this is source only annotation that is processed by annotation processor.

**Review by trentjeff — APPROVED** [review]

File: webserver/context-propagation/src/main/java/io/helidon/webserver/context/propagation/PropagationRecord.java (hunk 3 above)
Comment (tomas-langer) [reply webserver/context-propagation/src/main/java/io/helidon/webserver/context/propagation/PropagationRecord.java:72]: we already have a limit on maximal size of headers, so this should not be needed. We may want to add it on client side (so we do not have too big headers).

Comment (tomas-langer) [reply webserver/context-propagation/src/main/java/io/helidon/webserver/context/propagation/PropagationRecord.java:72]: I will leave this for our future selves - if there is a need to limit the size (or length of an array), let's get an issue and fix it

File: tests/integration/webserver/context-propagation/src/test/java/io/helidon/tests/integration/webserver/context/propagation/ContextPropagationTest.java (hunk 1 above)
Comment (tomas-langer) [reply tests/integration/webserver/context-propagation/src/test/java/io/helidon/tests/integration/webserver/context/propagation/ContextPropagationTest.java:67]: Added a negative test (part of the custom test method)

**Review by trentjeff — APPROVED** [review]

**Review by Verdent — APPROVED** [review]

**Pull request merged**